kroma-validator/challenger.go

```diff
@@ -998,65 +982,131 @@ func (c *Challenger) ProveFault(ctx context.Context, outputIndex *big.Int, chall
+// It sends requests to external RPC and immediately returns, then continuously checks the status of request and moves to next step.
+func (c *Challenger) proveFaultWithZkVm(
+	ctx context.Context, outputIndex *big.Int, challenge bindings.TypesChallenge, targetBlockNumber *big.Int, blockHash string, position *big.Int,
+) (*types.Transaction, bool, error) {
+	c.log.Info("crafting proveFaultWithZkVm tx", "outputIndex", outputIndex, "challenger", challenge.Challenger)
 
-		return c.colosseumContract.ProveFaultWithZkVm(txOpts, outputIndex, position, bindings.TypesZkVmProof{})
-	} else {
-		c.log.Info("crafting zkEVMProveFault tx", "outputIndex", outputIndex, "challenger", challenger)
+	l1Head := hex.EncodeToString(challenge.L1Head[:])
 
-		proof, err := c.PublicInputProof(ctx, blockNumber.Uint64())
-		if err != nil {
-			return nil, fmt.Errorf("failed to get public input proof(fault position blockNumber: %d): %w", blockNumber.Uint64(), err)
-		}
+	requestStatus, err := c.cfg.WitnessGenerator.RequestWitness(ctx, blockHash, l1Head)
+	if err != nil {
+		return nil, true, fmt.Errorf("failed to request witness(target block number: %s): %w", targetBlockNumber.String(), err)
+	}
+	if *requestStatus != chal.RequestCompleted {
+		c.log.Info("witness generation is not completed", "targetBlockNumber", targetBlockNumber, "status", *requestStatus)
+		return nil, true, nil
+	}
 
-		targetBlockNumber := new(big.Int).Add(blockNumber, common.Big1)
-		cCtx, cCancel := context.WithTimeout(ctx, c.cfg.NetworkTimeout)
-		defer cCancel()
-		trace, err := c.l2Client.GetBlockTraceByNumber(cCtx, targetBlockNumber)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get block trace(fault position blockNumber: %d): %w", targetBlockNumber.Uint64(), err)
-		}
+	witnessResult, err := c.cfg.WitnessGenerator.GetWitness(ctx, blockHash, l1Head)
+	if err != nil {
+		return nil, true, fmt.Errorf("failed to get witness(target block number: %s): %w", targetBlockNumber.String(), err)
+	}
 
-		traceBz, err := json.Marshal(trace)
-		if err != nil {
-			return nil, fmt.Errorf("failed to marshal block trace(fault position blockNumber: %d): %w", targetBlockNumber.Uint64(), err)
-		}
+	requestStatus, err = c.cfg.ZkVMProofFetcher.RequestProve(ctx, blockHash, l1Head, witnessResult.Witness)
+	if err != nil {
+		return nil, true, fmt.Errorf("failed to request proof(target block number: %s): %w", targetBlockNumber.String(), err)
+	}
+	if *requestStatus != chal.RequestCompleted {
+		c.log.Info("proof generation is not completed", "targetBlockNumber", targetBlockNumber, "status", *requestStatus)
+		return nil, true, nil
+	}
 
-		fetchResult, err := c.cfg.ProofFetcher.FetchProofAndPair(ctx, string(traceBz))
-		if err != nil {
-			return nil, fmt.Errorf("failed to fetch proof and pair(fault position blockNumber: %d): %w", targetBlockNumber.Uint64(), err)
-		}
+	proofResult, err := c.cfg.ZkVMProofFetcher.GetProof(ctx, blockHash, l1Head)
```

Comment:
Can't the zkVM prof response take a long time? 
If the proof fetch timeout occurs, it will repeat the request again from witness.

Reply:
No, now challenger does not have to wait for the prover response with maintaining HTTP connection. Since the prover JSON-RPC interface has been changed, challenger can just send request and then continuously fetch the status of request processing. You can see the reflected changes of challenger logic in the code!